File: haskell/step1_read_print.hs
import System.IO (hFlush, stdout)
import Control.Monad.Except (runExceptT)

import Readline (readline, load_history)
import Types
import Reader (read_str)
import Printer (_pr_str)

-- read

mal_read :: String -> IOThrows MalVal
mal_read = read_str

-- eval

eval :: MalVal -> MalVal
eval = id

-- print

mal_print :: MalVal -> String
mal_print = Printer._pr_str True

-- repl

rep :: String -> IOThrows String
rep line = mal_print <$> eval <$> mal_read line

repl_loop :: IO ()
repl_loop = do
    line <- readline "user> "
    case line of
        Nothing -> return ()
        Just "" -> repl_loop
        Just str -> do
            res <- runExceptT $ rep str
            out <- case res of
                Left mv -> return $ "Error: " ++ Printer._pr_str True mv
                Right val -> return val
            putStrLn out
            hFlush stdout
            repl_loop

main :: IO ()
main = do
    load_history

    repl_loop

File: dart/step1_read_print.dart
import 'dart:io';

import 'printer.dart' as printer;
import 'reader.dart' as reader;
import 'types.dart';

MalType READ(String x) => reader.read_str(x);

MalType EVAL(MalType x) => x;

String PRINT(MalType x) => printer.pr_str(x);

String rep(String x) {
  return PRINT(EVAL(READ(x)));
}

const prompt = 'user> ';
main() {
  while (true) {
    stdout.write(prompt);
    var input = stdin.readLineSync();
    if (input == null) return;
    var output;
    try {
      output = rep(input);
    } on reader.ParseException catch (e) {
      stdout.writeln("Error: '${e.message}'");
      continue;
    } on reader.NoInputException {
      continue;
    }
    stdout.writeln(output);
  }
}
